File: atcoder/abc/go/abc182/abc182_F/abc182_F.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
)

type scanner struct { sc *bufio.Scanner }
func newScanner(input io.Reader) *scanner {
	sc := bufio.NewScanner(input)
	sc.Split(bufio.ScanWords)
	sc.Buffer(make([]byte, 1024), int(1e+9))
	return &scanner{sc}	
}
var rdr = newScanner(os.Stdin)
const BUFSIZE = 10000000
var wrtr = bufio.NewWriterSize(os.Stdout, BUFSIZE)
func gs() string  { rdr.sc.Scan(); return rdr.sc.Text() }
func gi() int     { i,e := strconv.Atoi(gs()); if e != nil {panic(e)}; return i }
func gis(n int) []int  { res := make([]int,n); for i:=0;i<n;i++ { res[i] = gi() }; return res }
func gi2() (int,int) { return gi(),gi() }

type PI struct { x,y int }

func main() {
	//f1, _ := os.Create("cpu.prof"); pprof.StartCPUProfile(f1); defer pprof.StopCPUProfile()
    defer wrtr.Flush()
	infn := ""
	if infn == "" && len(os.Args) > 1 {	infn = os.Args[1] }
    if infn != "" {	f, e := os.Open(infn); if e != nil { panic(e) }; rdr = newScanner(f) }
	N,X := gi2()
	A := gis(N)
	cache := make(map[PI]int)
	var solveit func(int,int)int
	solveit = func(pos,rem int)int {
		v,ok := cache[PI{pos,rem}]; if ok { return v }
		ans := 0
		if pos == 0 || rem == 0 {
			ans = 1
		} else {
			absmax := 1_000_000_000_000_000_000
			if pos+1 < N  {absmax = A[pos+1]/A[pos] - 1 } 
			mincoins := rem / A[pos]
			residual := rem - rem / A[pos] * A[pos]
			if residual == 0 {
				ans = 1
			} else if mincoins == absmax {
				ans = solveit(pos-1,residual)
			} else {
				ans = solveit(pos-1,residual) + solveit(pos-1,A[pos]-residual)
			}
		}
		cache[PI{pos,rem}] = ans
		return ans
	}
	myans := solveit(N-1,X)
	fmt.Println(myans)
}

File: abc/python/abc182/abc182_F.py
import sys
infile = sys.stdin.buffer

def gs()  : return infile.readline().rstrip()
def gi()  : return int(gs())
def gf()  : return float(gs())
def gss() : return gs().split()
def gis() : return [int(x) for x in gss()]
def gfs() : return [float(x) for x in gss()]

def solveit(v,idx,N,A,cache) :
    if (v,idx) not in cache :
        if idx == 0 :
            cache[(v,idx)] = 1
        elif idx == N-1 :
            nn = v // A[idx]
            subtot = nn * A[idx]
            if subtot == v :
                cache[(v,idx)] = 1
            else :
                cache[(v,idx)] = solveit(v-subtot,idx-1,N,A,cache) + solveit(A[idx]-v+subtot,idx-1,N,A,cache)
                pass
        else :
            nn = v // A[idx]
            subtot = nn * A[idx]
            if subtot == v :
                cache[(v,idx)] = 1
            elif nn == A[idx+1] // A[idx] - 1 :
                cache[(v,idx)] = solveit(v-subtot,idx-1,N,A,cache)
            else :
                cache[(v,idx)] = solveit(v-subtot,idx-1,N,A,cache) + solveit(A[idx]-v+subtot,idx-1,N,A,cache)
    return cache[(v,idx)]

def main(infn="") :
    global infile
    infile = open(infn,"r") if infn else open(sys.argv[1],"r") if len(sys.argv) > 1 else sys.stdin
    N,X = gis()
    A = gis()
    cache = {}
    ans = solveit(X,N-1,N,A,cache)
    sys.stdout.write(str(ans)+'\n')

if __name__ == '__main__' :
    main()
    sys.stdout.flush()

